File: LeetCode/python3/503.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        stack = nums[::-1]
        ret = [-1 for i in range(len(nums))]
        for i in range(len(nums) - 1, -1, -1):
            while stack and stack[-1] <= nums[i]:
                stack.pop()
            if stack:
                ret[i] = stack[-1]
            stack.append(nums[i])
        return ret

File: LeetCode/cpp/503.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int> &nums) {
        int n = nums.size();
        stack<int> s;
        vector<int> ret(n, -1);
        for (int i = 0; i < n * 2; ++i) {
            while (!s.empty() && nums[s.top()] < nums[i % n]) {
                ret[s.top()] = nums[i % n];
                s.pop();
            }
            if (i < n)
                s.push(i);
        }
        return ret;
    }
};
